rj_gameplay/rj_gameplay/basic_play_selector.py

```diff
@@ -1,89 +1,271 @@
-    situations.MidfieldKickDirect: restart.DirectRestartPlay,
-    situations.GoalieClear: defensive_clear.DefensiveClear,
-    situations.Stop: defend_restart.DefendRestart,
-}
 
+POSSESS_MIN_DIST = 0.15
+MIN_PASS_SPEED = 0.9
+MIN_NEAR_BALL_DIST = 0.35
+
+
+class BallPos(Enum):
+    """Enum for representing the possession of the ball."""
+
+    OUR_BALL = 1
+    FREE_BALL = 2
+    THEIR_BALL = 3
 
-class BasicPlaySelector(situation.IPlaySelector):
-    """Play selector that returns a play based on the situation from analyzer.
 
-    Currently configured to take in only one play per situation. Situation to play mapping is determined in PLAY_DICT constant above. (This means using this class requires importing the whole file, rather than just the class.)
+class FieldLoc(Enum):
+    """Enum for representing where the ball is on the field."""
+
+    DEFEND_SIDE = 1
+    MIDFIELD = 2
+    ATTACK_SIDE = 3
+
+
+class HeuristicInformation:
+    """Class that represents all the heuristic information needed by the decision
+    tree situation analyzer.
+    """
+
+    __slots__ = ["ball_pos", "field_loc", "is_pileup"]
+
+    ball_pos: BallPos
+    field_loc: FieldLoc
+    is_pileup: bool
+
+    def __init__(self, world_state: stp.rc.WorldState, game_info: stp.rc.GameInfo):
+        self.field_loc = self.__calc_field_loc(world_state, game_info)
+        self.ball_pos = self.__calc_ball_pos(world_state, game_info)
+        self.is_pileup = self.__calc_pileup(world_state, game_info)
+
+    @staticmethod
+    def __calc_field_loc(
+        world_state: stp.rc.WorldState, game_info: stp.rc.GameInfo
+    ) -> FieldLoc:
+        """Computes the current location of the ball.
+        :param world_state:
+        :param game_info:
+        :return: The current FieldLoc.
+        """
+        ball_pos: np.ndarray = world_state.ball.pos
+
+        field_len: float = world_state.field.length_m
+        midfield: float = field_len / 2
+
+        if ball_pos[1] <= midfield:
+            return FieldLoc.DEFEND_SIDE
+        elif ball_pos[1] > midfield:
+            return FieldLoc.ATTACK_SIDE
+        else:
+            return FieldLoc.MIDFIELD
+
+    @staticmethod
+    def __calc_ball_pos(
+        world_state: stp.rc.WorldState, game_info: stp.rc.GameInfo
+    ) -> BallPos:
+        """Computes the current ball possession.
+        :param world_state:
+        :param game_info:
+        :return: The current BallPos.
+        """
+        for our_bot in world_state.our_robots:
+            if (
+                our_bot.has_ball_sense
+                or np.linalg.norm(
+                    np.array(world_state.ball.pos) - np.array(our_bot.pose[0:2])
+                )
+                < POSSESS_MIN_DIST
+            ):
+                return BallPos.OUR_BALL
+        for their_bot in world_state.their_robots:
+            if (
+                np.linalg.norm(
+                    np.array(world_state.ball.pos) - np.array(their_bot.pose[0:2])
```

Comment:
Maybe not necessary, but we could also take into account if the robot is facing the correct direction to be in possession of the ball. The robots turn reasonably fast, so maybe its not a big deal.

Reply:
yeah this should be done eventually